src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs

```diff
@@ -112,14 +112,63 @@ await InnerProcessMessages()
 
                     var nativeMessageId = Path.GetFileNameWithoutExtension(filePath).Replace(".metadata", "");
 
-                    var transaction = GetTransaction();
+                    await concurrencyLimiter.WaitAsync(cancellationToken)
+                        .ConfigureAwait(false);
 
-                    transaction.BeginTransaction(filePath);
+                    var transaction = GetTransaction();
 
-                    await concurrencyLimiter.WaitAsync(cancellationToken)
+                    var ableToLockFile = await transaction.BeginTransaction(filePath)
                         .ConfigureAwait(false);
 
-                    InnerProcessFile(transaction, nativeMessageId).Ignore();
+                    if (!ableToLockFile)
+                    {
+                        concurrencyLimiter.Release();
+                        continue;
+                    }
+
+                    ProcessFile(transaction, nativeMessageId)
+                        .ContinueWith(async t =>
+                        {
+                            try
+                            {
+                                if (t.Exception == null)
+                                {
+                                    receiveCircuitBreaker.Success();
+                                }
+                                else
+                                {
+                                    var baseEx = t.Exception.GetBaseException();
+
+                                    if (!(baseEx is OperationCanceledException))
+                                    {
+                                        await receiveCircuitBreaker.Failure(baseEx)
+                                            .ConfigureAwait(false);
+                                    }
+
+                                    transaction.Rollback();
+
+                                }
+
+                                var wasCommitted = await transaction.Complete()
+                                    .ConfigureAwait(false);
+
+                                if (wasCommitted)
+                                {
+                                    File.Delete(Path.Combine(bodyDir, nativeMessageId + BodyFileSuffix));
+                                }
+                            }
+                            catch (Exception ex)
+                            {
+                                await receiveCircuitBreaker.Failure(ex)
+                                    .ConfigureAwait(false);
+                            }
+                            finally
+                            {
+                                concurrencyLimiter.Release();
+                            }
+                        }, CancellationToken.None)
+                        .Unwrap()
```

Comment:
Because @danielmarbach said so :)

Pretty sure its because we want to `.Ignore()` the ProcessFile task and not the task returned from continuewith? ( Daniel?)